fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java

```diff
@@ -314,41 +313,57 @@ protected void computeCost(ExpressionInfo exprInfo) {
         Map<ColumnRefOperator, ScalarOperator> rightExpression = new HashMap<>();
-        if (onPredicates != null) {
-            ColumnRefSet useColumns = onPredicates.getUsedColumns();
-
-            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : expressionMap.entrySet()) {
-                if (!useColumns.contains(entry.getKey())) {
-                    continue;
-                }
-
-                if (entry.getValue().isConstantRef()) {
-                    // constant always on left
-                    leftExpression.put(entry.getKey(), entry.getValue());
-                    continue;
-                }
-
-                ColumnRefSet valueUseColumns = entry.getValue().getUsedColumns();
-                if (leftExprInfo.expr.getOutputColumns().containsAll(valueUseColumns)) {
-                    leftExpression.put(entry.getKey(), entry.getValue());
-                } else if (rightExprInfo.expr.getOutputColumns().containsAll(valueUseColumns)) {
-                    rightExpression.put(entry.getKey(), entry.getValue());
+        if (!onPredicates.isEmpty()) {
+            for (int i = 0; i < onPredicates.size(); i++) {
+                ScalarOperator predicate = onPredicates.get(i);
+                ColumnRefSet useColumns = predicate.getUsedColumns();
+                for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : expressionMap.entrySet()) {
+                    if (!useColumns.contains(entry.getKey())) {
+                        continue;
+                    }
+
+                    if (entry.getValue().isConstantRef()) {
+                        // constant always on left
+                        leftExpression.put(entry.getKey(), entry.getValue());
+                        continue;
+                    }
+
+                    ColumnRefSet valueUseColumns = entry.getValue().getUsedColumns();
+                    if (leftExprInfo.expr.getOutputColumns().containsAll(valueUseColumns)) {
+                        leftExpression.put(entry.getKey(), entry.getValue());
+                        continue;
+                    } else if (rightExprInfo.expr.getOutputColumns().containsAll(valueUseColumns)) {
+                        rightExpression.put(entry.getKey(), entry.getValue());
+                        continue;
+                    }
+
+                    ColumnRefSet allChildColumns = new ColumnRefSet();
+                    allChildColumns.union(leftExprInfo.expr.getOutputColumns());
+                    allChildColumns.union(rightExprInfo.expr.getOutputColumns());
+                    if (allChildColumns.containsAll(valueUseColumns)) {
+                        // depend on two children, must rewrite to origin expression
+                        ReplaceColumnRefRewriter rewriter =
+                                new ReplaceColumnRefRewriter(Map.of(entry.getKey(), entry.getValue()));
+                        predicate = rewriter.rewrite(predicate);
+                        onPredicates.set(i, predicate);
```

Comment:
Stupid question, why this can improve JoinReorder's preformance?

Reply:
This removes the final plan validation step.